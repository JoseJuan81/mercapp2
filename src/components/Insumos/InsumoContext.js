import { createContext } from 'react';

export const InsumoContext = createContext(null);