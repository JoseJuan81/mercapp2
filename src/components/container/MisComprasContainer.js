import React from 'react';

import { PaginaResumenCompras } from '../../Pages/PaginaResumenCompras';
import { ResumenMenuMobile } from '../Menu/ResumenMenuMobile';

export const MisComprasContainer = () => {
    return (
        <div
            className="layout__page"
        >
            <PaginaResumenCompras />

            <ResumenMenuMobile />
        </div>
    )
}
