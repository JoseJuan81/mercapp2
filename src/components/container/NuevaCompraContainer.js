import React from 'react';

import { PaginaNuevaCompra } from '../../Pages/PaginaNuevaCompra';
import { NuevaCompraMenuMobile } from '../Menus/NuevaCompraMenuMobile';

export const NuevaCompraContainer = () => {
    return (
        <div
            className="layout__page"
        >
            <PaginaNuevaCompra />

            <NuevaCompraMenuMobile />

        </div>
    )
}
