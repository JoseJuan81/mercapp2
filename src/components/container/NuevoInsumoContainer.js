import React from 'react';


import { PaginaNuevoInsumo } from '../../Pages/PaginaNuevoInsumo';

export const NuevoInsumoContainer = () => {

    return (
        <div
            className="layout__page"
        >

            <PaginaNuevoInsumo />

        </div>
    )
}
