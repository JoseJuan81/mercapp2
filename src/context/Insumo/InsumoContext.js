import { createContext } from 'react';

/**
 * Crear contexto de Insumo
 */
export const InsumoContext = createContext(null);
