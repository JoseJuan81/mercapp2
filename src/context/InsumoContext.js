import { createContext, useEffect, useMemo, useState } from 'react';
import { removeItemFromArrayByProp } from 'functionallibrary';

import { useIndexDB } from '../hooks/useIndexDB';

import { getFromLocalStorage, setInLocalStorage } from '../helper/localStorage';
import {
    initialInsumos,
    onFilter,
    onSearch,
    updateItemInArrayById,
    selectAllInsumos,
    unSelectAllInsumos,
    updateTotal,
    updatingQuantity,
} from '../helper/helperInsumoContext';

/**
 * Crear contexto de Insumo
 */
export const InsumoContext = createContext(null);

const dataSelected = getFromLocalStorage('selected-insumos') || [];

/**
 * @description funcion wrapper del contexto
 * @param {ReactComponent} children
 */
export const InsumoStore = ({ children }) => {

    // obtener insumos de indexDB
    const {
        deleteInsumo,
        insumos: insumosInLocalBD,
        setNewInsumo
    } = useIndexDB();


    // Ordenar insumos provenientes de indexDB y actualizarlos con los seleccionados
    const initInsumos = useMemo( () => initialInsumos( insumosInLocalBD ), [insumosInLocalBD] );

    const [insumos, setInsumos] = useState(initInsumos);
    const [selectedInsumos, setSelectedInsumos] = useState(dataSelected);
    const [total, setTotal] = useState(0);


    /**
     * @description: seleccionar o deseleccionar un insumo
     * @param {object} newInsumo - insumo
     */
    const toogleCheck = (newInsumo) => {

        // Actualizar insumos
        setInsumos(
            updateItemInArrayById(insumos, newInsumo)
        )

        // actualizar insumos seleccionados
        const updatedSelectedInsumos = newInsumo.checked
            ? [newInsumo, ...selectedInsumos]
            : removeItemFromArrayByProp('id', newInsumo.id, selectedInsumos);
        
        setSelectedInsumos( [...updatedSelectedInsumos] );
        setInLocalStorage('selected-insumos', updatedSelectedInsumos);


    }

    /**
     * @description: actualiza cantidad del insumo
     * @param {string} id 
     * @param {number} quantity 
     */
    const updateQuantityInSelectedInsumo = (id, quantity) => {

        setSelectedInsumos(
            updatingQuantity( id, quantity, selectedInsumos )
        );

    }

    /**
     * @description: selecciona todos los insumos
     * @param {array} insumos - arreglo de insumos
     */
    const selectingAllInsumos = (insumos) => {

        const allSelected = selectAllInsumos(insumos);

        setInsumos( allSelected );
        setSelectedInsumos([...allSelected]);
        setInLocalStorage('selected-insumos', allSelected);
    }
    
    /**
     * @description: deselecciona todos los insumos
     * @param {array} insumos - arreglo de insumos
     */
    const unSelectingAllInsumos = () => {
        
        const allUnSelected = unSelectAllInsumos( initialInsumos() );

        setInsumos( allUnSelected );
        setSelectedInsumos( [] );
        setInLocalStorage( 'selected-insumos', [] );
    };

    /**
     * @description: Busca insumos usando el titulo
     * @param {string} searchVal - criterio de busqueda
     */
    const searchingInsumos = (searchVal) => {

        if(searchVal === '' || searchVal === null || searchVal === undefined) {

            setInsumos( initialInsumos() );
        } else {

            const result = onSearch( insumos, searchVal );
            setInsumos( result );
        }

    }

    /**
     * @description: Busca insumos usando los labels
     * @param {string} searchVal - criterio de buscqueda
     */
    const filteringInsumos = (searchVal) => {

        if(searchVal === '' || searchVal === null || searchVal === undefined) {

            setInsumos( initialInsumos() );
        } else {

            const result = onFilter( insumos, searchVal );
            setInsumos( result );
        }

    }

    const addingNewInsumo = async (newInsumo) => {

        try {

            await setNewInsumo( newInsumo );
            setInsumos( i => [{ ...newInsumo }, ...i] );

        } catch(err) {
            console.log('Erro al crear insumo', err);
        }

    }

    const deletingInsumo = async (insumoId) => {

        await deleteInsumo(insumoId);

        setInsumos(
            removeItemFromArrayByProp('id', insumoId, insumos)
        );
    }

    // Total de insumos seleccionados => price * quantity
    useEffect( () => {

        setTotal(
            updateTotal( selectedInsumos )
        );
        
    }, [selectedInsumos])

    useEffect( () => {

        setInsumos(initInsumos);

    }, [initInsumos])

    const insumoContextProps = {
        addingNewInsumo,
        deletingInsumo,
        selectingAllInsumos,
        filteringInsumos,
        insumos,
        unSelectingAllInsumos,
        searchingInsumos,
        selectedInsumos,
        total,
        toogleCheck,
        updateQuantityInSelectedInsumo,
    }

    return (
        <InsumoContext.Provider
            value={ insumoContextProps }
        >
            { children }

        </InsumoContext.Provider>
    )
}