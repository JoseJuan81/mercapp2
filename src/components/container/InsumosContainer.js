import React, { useCallback, useContext, useEffect, useState } from 'react';

import { ListadoInsumos } from '../Insumos/ListadoInsumos';
import { Searcher } from '../Genericos/form/Searcher';
import { InsumosMenuMobile } from '../Insumos/InsumosMenuMobile';

import { BottomModal } from '../Genericos/modal/BottomModal';
import { InsumoForm } from '../Genericos/form/InsumoForm';
import { InsumoContext } from '../../context/Insumo/InsumoContext';
import { isEmpty } from 'functionallibrary';


export const InsumosContainer = () => {
    // console.log('3 INSUMO CONTAINER');

    // Contexto de Insumo
    const {
        filteringInsumos,
        insumoToUpdate,
        searchingInsumos,
        setInsumoToUpdate,
    } = useContext( InsumoContext );

    // variables para mostrar u ocultar buscadores
    const [showSearch, setShowSearch] = useState(false);
    const [showFilter, setShowFilter] = useState(false);

    // variables para mostrar u ocular modal
    const [showModal, setShowModal] = useState(false);

    // accion para cambiar valor de variables para mostrar u ocultar buscadores y modal
    const toogleShowSearch = useCallback( () => {

        setShowFilter(() => false);
        setShowSearch(s => !s);

    }, [setShowFilter, setShowSearch]);

    // accion para cambiar valor de variables para mostrar u ocultar buscadores y modal
    const toogleShowFilter = useCallback( () => {

        setShowSearch(() => false);
        setShowFilter(s => !s);

    }, [setShowSearch, setShowFilter]);

    const openModal = useCallback( () => setShowModal( true ), [setShowModal]);
    const closeModal = useCallback( () => {

        setShowModal( false );
        setInsumoToUpdate();
    }, [setShowModal]);

    useEffect( () => {

        setShowModal( !isEmpty( insumoToUpdate ) );
    }, [insumoToUpdate]);

    return (
        <>
            <div
                className={`
                    flex
                    mb-4
                    overflow-hidden
                    ${showSearch || showFilter ? 'max-h-20' : 'max-h-0'}
                    duration-300
                `}
            >
                {
                    showSearch && <Searcher
                        onSearch={ searchingInsumos }
                        placeholder="Buscar insumo"
                    />
                }
                {
                    showFilter && <Searcher
                        onSearch={ filteringInsumos }
                        placeholder="Filtrar insumo"
                    />
                }

            </div>

            <ListadoInsumos />

            <InsumosMenuMobile
                toogleShowSearch={ toogleShowSearch }
                toogleShowFilter={ toogleShowFilter }
                openModal={ openModal }
            />


            { showModal &&
            
                <BottomModal show={ showModal } >

                    <InsumoForm closeModal={ closeModal } />

                </BottomModal>

            }

        </>
    )
}
