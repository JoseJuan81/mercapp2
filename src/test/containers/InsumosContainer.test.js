import { act } from '@testing-library/react';
import { mount } from 'enzyme';
import { InsumosContainer } from "../../components/container/InsumosContainer"
import { InsumoContext } from '../../context/InsumoContext';


const selectingAllInsumos = jest.fn();
const unSelectingAllInsumos = jest.fn();
const selectedInsumos = [];
const total = 0;
const toogleCheck = jest.fn();
const updateQuantityInSelectedInsumo = jest.fn();

const fakeInsumo = {
    id: '23f8e315e5b27',
    title: "Aceite de Oliva",
    price: 20,
    currency: "S/.",
    labels: ["500ml", "Carbonell", "metro", "2 und"]
}

describe('Pruebas sobre archivo InsumosContainer.js', () => {

    const wrapper = mount(
        <InsumoContext.Provider
            value={{
                selectingAllInsumos,
                insumos: [fakeInsumo],
                unSelectingAllInsumos,
                selectedInsumos,
                total,
                toogleCheck,
                updateQuantityInSelectedInsumo
            }}
        >

            <InsumosContainer />

        </InsumoContext.Provider>
    );

    test('se debe mostrar correctamente <InsumosContainer />', () => {

        expect( wrapper ).toMatchSnapshot();
    });

    test('Se debe llamar toogleCheck al hacer click en un insumo', () => {

        const firstLi = wrapper.find('li').at(0);

        act( () => {

            firstLi.prop('onClick')(fakeInsumo);

        })
        
        expect( toogleCheck ).toHaveBeenCalled();
    })
})