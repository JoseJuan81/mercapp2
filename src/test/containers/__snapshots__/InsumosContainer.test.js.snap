// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas sobre archivo InsumosContainer.js se debe mostrar correctamente <InsumosContainer /> 1`] = `
Array [
  <div
    className="
                    flex
                    mb-4
                    overflow-hidden
                    max-h-0
                    duration-300
                "
  />,
  <ul
    className=""
  >
    <li
      className="mb-3"
      key="Aceite de Oliva-0-function random() { [native code] }"
      onClick={[Function]}
    >
      <div
        className="
                duration-200
                rounded-lg undefined
                border border-solid border-warmGray-300
            "
      >
        <dl
          className="flex p-2 overflow-hidden relative"
        >
          <div
            className="flex flex-auto"
          >
            <dt
              className="
                duration-200
                flex-auto
                text-2xl font-bold text-warmGray-400
                pr-6
            "
            >
              Aceite de Oliva
            </dt>
          </div>
          <div
            className="
                absolute right-0
                pr-2
                bg-warmGray-100
                flex
                transform translate-x-24
                border border-solid border-white
                rounded-l-full
                duration-300
            "
          >
            <button
              className="
                    bg-white
                    icon-circle-left
                    flex items-center justify-center
                    text-warmGray-500
                    text-2xl
                    transform rotate-0
                    w-10
                    false
                    border border-solid border-white
                    duration-300
                "
              onClick={[Function]}
              type="button"
            />
            <div
              className=" grid grid-cols-2 gap-2 "
            >
              <button
                className=" icon-pencil px-2 text-2xl text-blue-500 "
              />
              <button
                className=" icon-bin2 px-2 text-2xl text-rose-500 "
              />
            </div>
          </div>
        </dl>
        <div
          className="
                        rounded-br-lg rounded-bl-lg
                        bg-warmGray-100
                        p-1
                    "
        >
          <ul
            className="flex flex-wrap"
          >
            <li
              className="
                            bg-white
                            text-xs text-warmGray-800
                            m-1 p-1
                        "
              key="500ml-0-function random() { [native code] }"
            >
              500ml
            </li>
            <li
              className="
                            bg-white
                            text-xs text-warmGray-800
                            m-1 p-1
                        "
              key="Carbonell-1-function random() { [native code] }"
            >
              Carbonell
            </li>
            <li
              className="
                            bg-white
                            text-xs text-warmGray-800
                            m-1 p-1
                        "
              key="metro-2-function random() { [native code] }"
            >
              metro
            </li>
            <li
              className="
                            bg-white
                            text-xs text-warmGray-800
                            m-1 p-1
                        "
              key="2 und-3-function random() { [native code] }"
            >
              2 und
            </li>
          </ul>
        </div>
      </div>
    </li>
  </ul>,
  <div
    className=" grid grid-cols-5 fixed bottom-0 left-0 w-full h-16 bg-white border-t border-solid border-warmGray-300 shadow-inner "
  >
    <button
      className="icon-checkmark text-lime-400 text-3xl"
      onClick={[Function]}
      title="Seleccionar todo"
    />
    <button
      className=" icon-checkmark text-3xl text-warmGray-300 "
      onClick={[Function]}
      title="Deseleccionar todo"
    />
    <button
      className="icon-search text-3xl text-rose-500"
      onClick={[Function]}
      title="Buscar"
    />
    <button
      className="icon-filter text-3xl text-blue-400"
      onClick={[Function]}
      title="Filtrar"
    />
    <button
      className=" icon-plus text-3xl text-lime-600 "
      onClick={[Function]}
      title="Agregar"
    />
  </div>,
]
`;
